### PR TITLE
Feat(eos_designs): Support different ipv4 pool for mlag ibgp peerings

### LINKED ISSUE #1797: Ability to use different ip address for ibgp peering in non default vrf
### Enhancement summary

AVD today defaults to using mlag_peer_l3_ipv4_pool to create the bgp peering. 
In the vxlan EVPN environment, ibgp peering created over the SVI in a non default vrf requires the ability to use different ip address across different vrfs. 

### Which component of AVD is impacted

eos_designs

### Use case example

VXLAN evpn deployment which is acting as aggregation for the different routes for running the overlay SDWAN solution. 
Although reusing the ip addressing across different vrfs should work, customer prefers to use different ip addressing for the ibgp peering vlans.

### Describe the solution you would like

1. SVI creation with custom ip pool
2. iBGP Peering configured using the ip address in the /31 subnet assigned to the SVI created in step 1
3. IBGP peer activation in address family IPv4 for that non default vrf where ibgp peering is created
4. Ability to assign it to different peer group (by default ibgp peer is getting assigned to peer group MLAG-IPv4-UNDERLAY-PEER)

### Describe alternatives you have considered

Nothing found yet.

### Additional context

_No response_

### Contributing Guide

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by tgodaA**
## Change Summary

Support to add user defined ip for mlag ibgp peering.

## Related Issue(s)

Fixes #1797 

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes

Add knob (mlag_ibgp_peering_ipv4_pool) which works on both new and old data-models:

1) Old
```yaml
tenants:
  <tenant1>:
    vrfs:
      <vrf1>:
        mlag_ibgp_peering_ipv4_pool: < IPv4address/Mask | default -> mlag_l3_ip -> mlag_ip >
```

2) New
```yaml
tenants:
  - name: < tenant1 >
    vrfs:
      - name: <vrf1>
        mlag_ibgp_peering_ipv4_pool: < IPv4address/Mask | default -> mlag_l3_ip -> mlag_ip >
```

## How to test

molecule converge -s eos_designs_unit_tests
molecule converge -s eos_designs_unit_testsv4.0

Optional: make refresh-facts

### User Checklist

- [x] eos_designs_unit_tests and eos_designs_unit_testsv4.0 artifacts should be same

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2 [hunk 1]
```diff
@@ -56,7 +56,10 @@
                                                          tenant.enable_mlag_ibgp_peering_vrfs,
                                                          true) %}
 {%                         if configure_mlag_ibgp_peering %}
-        {{ switch.mlag_peer_l3_ip | arista.avd.default(switch.mlag_peer_ip) }}:
+{%                             if vrf.mlag_ibgp_peering_ipv4_pool is arista.avd.defined %}
+{%                                 set mlag_ibgp_peering_peer_ip = vrf.mlag_ibgp_peering_ipv4_pool | ansible.netcommon.ipaddr('subnet') | ansible.netcommon.ipmath(switch.mlag_peer_id - 1) %}
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2]: Any reason for which the formula is not the same as in the other mlag_ip templates:
https://github.com/aristanetworks/ansible-avd/blob/707ca25feb8ee6e5a5f2d87c1eb6bb71dd10e0f6/ansible_collections/arista/avd/roles/eos_designs/templates/ip_addressing/mlag-l3-ip-primary.j2#L1
?

And another question - would it be possible to compute this as a fact maybe?

File: ansible_collections/arista/avd/roles/eos_designs/doc/network-services-v4.0.md [hunk 2]
```diff
@@ -144,6 +144,11 @@ svi_profiles:
         # "vrf_id" is preferred over "vrf_vni" for MLAG IBGP peering vlan, see "mlag_ibgp_peering_vrfs.base_vlan" for details
         vrf_id: < 1-1024 >
 
+        # MLAG IBGP Peering IPv4 Pool | Optional
+        # Uses the switch.id and switch.mlag_peer_id to calculate the mlag_ibgp_peering_ip and mlag_ibgp_peering_peer_ip
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/network-services-v4.0.md]: these switch.id and switch.mlag_peer_id are not known to the user, so I suggest rewording to
```
# The subnet used for iBGP peering in the VRF.
# Each MLAG pair will be assigned a subnet based on the ID of the primary MLAG switch
# If not set, "mlag_peer_l3_ipv4_pool" or "mlag_peer_ipv4_pool" will be used

File: ansible_collections/arista/avd/roles/eos_designs/doc/network-services-v4.0.md [hunk 3]
```diff
@@ -144,6 +144,11 @@ svi_profiles:
...
+        # mlag_ibgp_peering_ip and mlag_ibgp_peering_peer_ip default -> mlag_l3_ip, mlag_peer_l3_ip -> mlag_ip, mlag_peer_ip
+        mlag_ibgp_peering_ipv4_pool: < IPv4_address/Mask >
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/network-services-v4.0.md:151]: In a multi-pod setup, it would probably make sense to use a pool per POD. Check "loopback_ip_pools". Should we change to such a model, or keep this for now, and add the other one in parallel if needed?

File: ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2 (hunk 1 above)
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2]: so this is not allowing a custom template to be supplied like we have for the general mlag peering.
The logic here is not the same as we have for the general mlag peering. You should use primary_id on both, and -1 on IP only for primary. This way we don't require the mlag peers to have consecutive IDs.

**Review by ClausHolbechArista — DISMISSED** [review]

File: ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2 (hunk 1 above)
Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2]: This is set on Tenant networks for every VRF individually based on switch id. So I don't think of it as a switch fact. I can adjust the code to make it work like a switch fact but that wasn't intended.

Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2]: Correct. Will change the code.

Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2]: With respect to the formula, I derived it using both id and peer_id, which I need to change. That's why the formula is different.

Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2]: But the only problem that I see was with the user definition. For example: switch pairs with ids (1,3) and (2,4). The switch 3 and switch 2 will end up having same IP.

I still agree with your suggestion as taking id and peer_id to calculate the IPs will not keep the p2p links in the same subnet (/31). For example (1,3) -> .0/31 and .2/31 are not in the same subnet.

Another problem with non-consecutive switch ids is: **The ip and peer_ip** might end up in a different /31 subnets. For example the switch pair (2,4) -> .1/31 and .2/31 (different subnets)

Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2]: Probably I need a stronger algorithm to solve all these issues.

Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2]: @ClausHolbechArista,  I derived a new algorithm, which will make sure all of the above criteria is addressed.

1) Calculating the new IP_address only based on primary_id (let's say **id**)
2) **ip_suffix** = id - (id % 2)
3) For a secondary node: **ip_suffix** = ip_suffix + 1

Example:
For switch pair (1,2) ids: 
1) id = 1
2) ip_suffix = 1- 1%2 = 0
3) for primary switch: sw1, ip_suffix = 0, secondary switch: sw3, ip_suffix = 1
So **.0/31** and **.1/31** are provided to both the switches

Switch pair (2,4):
1) id = 2
2) ip_suffix = 2 - 2%2 = 2
3) for primary switch: sw2, ip_suffix = 2, secondary switch: sw4, ip_suffix = 3
So **.2/31** and **.3/31** are provided to both the switches

In above both cases, there is no overlap in IPs.
Also, I could have gone with: ```{{ switch_data.combined.mlag_peer_l3_ipv4_pool | ansible.netcommon.ipaddr('network') | ansible.netcommon.ipmath((mlag_primary_id - 1) * 2) }} ``` the mlag_ip derivation but here, some IPs are wasted.

For example sw(1,2) -> use **.0/31 .1/31** and sw(3,4) -> use **.4/31 .5/31**, where as my new formula will be using the IPs sw(1,2) -> use **.0/31 .1/31** and sw(3,4) -> use **.2/31 .3/31**. Less IP wastage and more efficiency.

Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2]: @gmuloc here is the formula that I wanted to use (with respect to your above comment).

File: ansible_collections/arista/avd/roles/eos_designs/doc/network-services-v4.0.md (hunk 3 above)
Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/doc/network-services-v4.0.md:151]: Wanted to implement it per vrf.

File: ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2 (hunk 1 above)
Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2]: Using the same formula.

Comment (gmuloc) [reply ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2]: @tgodaA:

if I have two pairs of ids: A = (2,4) and B = (3,5)

it would lead to 
A:
1. id = 2
2. primary_ip_suffix = 2 - 2%2 = 2
3. secondary_ip_suffix = primary_ip_suffix + 1 = 3


B: 
1. id = 3
2. primary_ip_suffix = 3 - 3%2 = 2
3. secondary_ip_suffix = primary_ip_suffix + 1 = 3

so there would be a clash

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3A.cfg [hunk 4]
```diff
@@ -652,7 +652,7 @@ interface Vlan3010
    no shutdown
    mtu 1500
    vrf Tenant_A_WEB_Zone
-   ip address 10.255.251.6/31
+   ip address 172.31.11.6/31
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3A.cfg:655]: This IP is the same as DC1-SVC3B on the same VLAN - I don't think that is what is expected

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3B.cfg [hunk 5]
```diff
@@ -627,7 +627,7 @@ interface Vlan3010
    no shutdown
    mtu 1500
    vrf Tenant_A_WEB_Zone
-   ip address 10.255.251.7/31
+   ip address 172.31.11.6/31
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3B.cfg]: It should be
```suggestion
   ip address 172.31.11.7/31
```
Otherwise it is the same as the other end

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3A.cfg [hunk 6]
```diff
@@ -961,7 +961,7 @@ router bgp 65103
       route-target import evpn 11:11
       route-target export evpn 11:11
       router-id 192.168.255.12
-      neighbor 10.255.251.7 peer group MLAG-PEERS
+      neighbor 172.31.11.8 peer group MLAG-PEERS
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3A.cfg]: Should not it be 
```suggestion
      neighbor 172.31.11.7 peer group MLAG-PEERS
```

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3B.cfg [hunk 7]
```diff
@@ -935,7 +935,7 @@ router bgp 65103
       route-target import evpn 11:11
       route-target export evpn 11:11
       router-id 192.168.255.13
-      neighbor 10.255.251.6 peer group MLAG-PEERS
+      neighbor 172.31.11.8 peer group MLAG-PEERS
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3B.cfg]: Should not it be 
```suggestion
      neighbor 172.31.11.6 peer group MLAG-PEERS
```

File: ansible_collections/arista/avd/roles/eos_designs/templates/ip_addressing/mlag-ibgp-peering-ip-secondary.j2 [hunk 8]
```diff
@@ -0,0 +1 @@
+{% set mlag_ibgp_peering_ip.peer_ip = vrf.mlag_ibgp_peering_ipv4_pool | ansible.netcommon.ipaddr('subnet') | ansible.netcommon.ipmath((switch.mlag_switch_ids.secondary - 1) * 2) %}
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/templates/ip_addressing/mlag-ibgp-peering-ip-secondary.j2]: That's not the same formula as the one in the [current secondary templates](https://github.com/aristanetworks/ansible-avd/blob/devel/ansible_collections/arista/avd/roles/eos_designs/templates/ip_addressing/mlag-ip-secondary.j2) and it is leading to having a wrong IP set for the remote neighbor

```suggestion
{% set mlag_ibgp_peering_ip.peer_ip = vrf.mlag_ibgp_peering_ipv4_pool | ansible.netcommon.ipaddr('subnet') | ansible.netcommon.ipmath(((switch.mlag_switch_ids.primary - 1) * 2) + 1) %}
```

File: ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2 [hunk 9]
```diff
@@ -67,7 +67,11 @@
 {%                     endif %}
       neighbors:
 {%                     if configure_mlag_ibgp_peering %}
-        {{ switch.mlag_peer_l3_ip | arista.avd.default(switch.mlag_peer_ip) }}:
+{%                         if vrf.mlag_ibgp_peering_ipv4_pool is arista.avd.defined %}
+{%                             set mlag_ibgp_peering_ip = namespace() %}
+{%                             include switch.ip_addressing.mlag_ibgp_peering_ip_secondary %}
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2]: this is assuming that the current peer is always primary

File: ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2 [hunk 10]
```diff
@@ -67,7 +67,11 @@
...
+{%                         endif %}
+        {{ mlag_ibgp_peering_ip.peer_ip | arista.avd.default(switch.mlag_peer_l3_ip, switch.mlag_peer_ip) }}:
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/templates/network_services/router-bgp-vrfs.j2]: `mlag_ibgp_peering_ip.peer_ip ` will always have the same value whether it is primary or secondary switch we are working on (cf generated configs)

File: ansible_collections/arista/avd/roles/eos_designs/templates/network_services/vlan-interfaces.j2 [hunk 11]
```diff
@@ -136,6 +136,10 @@ vlan_interfaces:
 {%             if configure_mlag_ibgp_peering %}
 {%                 set vrf_id_int = vrf.vrf_id | arista.avd.default(vrf.vrf_vni) | int %}
 {%                 if vrf.name != 'default' %}
+{%                     if vrf.mlag_ibgp_peering_ipv4_pool is arista.avd.defined %}
+{%                         set mlag_ibgp_peering_ip = namespace() %}
+{%                         include switch.ip_addressing.mlag_ibgp_peering_ip_primary %}
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/templates/network_services/vlan-interfaces.j2]: This is assuming the switch we are working on is always primary

File: ansible_collections/arista/avd/roles/eos_designs/templates/network_services/vlan-interfaces.j2 [hunk 12]
```diff
@@ -145,7 +149,7 @@ vlan_interfaces:
 {%                     if underlay_rfc5549 is arista.avd.defined(true) and overlay_mlag_rfc5549 is arista.avd.defined(true) %}
     ipv6_enable: true
 {%                     else %}
-    ip_address: {{ switch.mlag_l3_ip | arista.avd.default(switch.mlag_ip) }}/31
+    ip_address: {{ mlag_ibgp_peering_ip.ip | arista.avd.default(switch.mlag_l3_ip, switch.mlag_ip) }}/31
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/templates/network_services/vlan-interfaces.j2]: `mlag_ibgp_peering_ip.ip ` will always have the same value (cf generated config)

**Review by gmuloc — CHANGES_REQUESTED** [review]

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3B.cfg (hunk 7 above)
Comment (tgodaA) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/DC1-SVC3B.cfg]: My bad. I need to change the logic. I have been using the same wrong logic again and again.

File: ansible_collections/arista/avd/roles/eos_designs/templates/ip_addressing/mlag-ibgp-peering-ip-primary.j2 [hunk 13]
```diff
@@ -0,0 +1 @@
+{% set mlag_ibgp_peering_ip.ip = vrf.mlag_ibgp_peering_ipv4_pool | ansible.netcommon.ipaddr('subnet') | ansible.netcommon.ipmath((switch.mlag_switch_ids.primary - 1) * 2) %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/templates/ip_addressing/mlag-ibgp-peering-ip-primary.j2]: These templates ususally returns a string. I would prefer if we could keep the same style.

**Comment by ClausHolbechArista** [comment]
Moving to draft. Mark as ready for review when logic has been fixed.

File: ansible_collections/arista/avd/roles/eos_designs/templates/ip_addressing/mlag-ibgp-peering-ip-primary.j2 (hunk 13 above)
Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/ip_addressing/mlag-ibgp-peering-ip-primary.j2]: I don't see a way to return it that way.

**Review by lermilov — APPROVED** [review]
LGTM!

**Review by gmuloc — APPROVED** [review]
LGTM

File: ansible_collections/arista/avd/roles/eos_designs/templates/ip_addressing/mlag-ibgp-peering-ip-primary.j2 (hunk 13 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/ip_addressing/mlag-ibgp-peering-ip-primary.j2]: Then we should change it to not use a template, and then provide an `offset` variable, to address any special assignment needs.

**Review by ClausHolbechArista — APPROVED** [review]
LGTM
